src/main/haskell/kore/src/Kore/Debug.hs

```diff
@@ -60,42 +61,117 @@ module Kore.Debug
     , traceNonErrorMonad
     , traceFunction
     , applyWhen
+    , debugArg
+    , DebugPlace (..)
+    , DebugResult (..)
     ) where
 
-import Control.Monad.Trans.Except
-import Control.Monad.Trans.Maybe
-import Debug.Trace
+import           Control.Monad.Trans.Except
+import           Control.Monad.Trans.Maybe
+import           Data.List
+                 ( intercalate )
+import           Data.Map
+                 ( Map )
+import qualified Data.Map as Map
+import           Debug.Trace
+
+{-| Identifiers for places where we're doing debug.
+
+The intent id to use D_Generic for places where we're adding temporary debug
+statements and the others for permanent debug statements
+-}
+data DebugPlace
+    = D_Generic !String !DebugResult
+    | D_OnePath_Step_transitionRule
+    | D_BaseStep_stepWithRule
+    | D_Function_evaluateApplication
+  deriving (Eq, Ord, Show)
+
+data DebugArg = DebugArg { name :: !String, value :: !String }
+
+instance Show DebugArg where
+    show DebugArg {name, value} = name ++ "=" ++ show value
+
+{-| Whether to dispay the function/action result when the function ends.
+-}
+data DebugResult = DebugResult | DebugNoResult
+  deriving (Eq, Ord, Show)
+
+{-| Wraps a field in order to be displayed when debugging -}
+debugArg :: Show a => String -> a -> DebugArg
+debugArg n v = DebugArg {name = n, value = show v}
 
 {-|Applies a function only when the condition is met.  Useful for conditional
 debugging, among other things.
 -}
 applyWhen :: Bool -> (a -> a) -> (a -> a)
 applyWhen b f = if b then f else id
 
+{-| Maps debug places to their debug settings.
+
+If a place other than `D_Generic` is missing from this map, we will not log
+debugging information for that place.
+
+Example:
+
+enabledPlaces = onePathWithFunctionNames
+
+-}
+enabledPlaces :: Map DebugPlace DebugResult
+enabledPlaces = Map.empty
+
+onePathWithFunctionNames :: Map DebugPlace DebugResult
+onePathWithFunctionNames =
+    Map.insert D_Function_evaluateApplication DebugNoResult
+    $ Map.insert D_OnePath_Step_transitionRule DebugResult
+    $ Map.singleton D_BaseStep_stepWithRule DebugResult
+
+traceWhenEnabled
+    :: DebugPlace -> (DebugResult -> a -> a) -> (a -> a)
+traceWhenEnabled place logger =
+    case place of
+        D_Generic _ debugResult -> logger debugResult
+        _ -> case Map.lookup place enabledPlaces of
+            Nothing -> id
+            Just debugResult -> logger debugResult
+
 {-|Wraps an 'ExceptT' action for printing debug messages, similar to 'trace'.
 
 It prints the name and the start values before the action, and the action
 result after.
 -}
 traceExceptT
```

Comment:
We should make sure the `trace*` functions have the `HasCallStack` constraint so that the call stack of the child is passed through properly.

Reply:
Would we get benefits right now, or we would need to do something else to use child stack traces?

Reply:
As written, any function wrapped with `trace...` that calls `error` will produce a call stack that ends at `trace...`. Depending on the error, that may capture enough information, or it may not.

Reply:
Ah, I see, thanks!